gcs_light_client.mediawiki

```diff
@@ -0,0 +1,1056 @@
+| [32]byte
+| Hash to stop at.
+|-
+| 1
+| FilterType
+| byte
+| Type of filter header being requested.
+|}
+
+The <code>BlockLocators</code> within the message are to be interpreted
+identically to the <code>BlockLocators</code> within Bitcoin's
+<code>getheaders</code> and <code>getblocks</code> messages <ref>https://en.bitcoin.it/wiki/Protocol_documentation</ref>.
+
+The <code>cfheaders></code> message MUST be sent in response to a
+<code>getcfheaders</code> message for a particular block hash. The
+<code>cfheaders</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Descriptions
+! Data Type
+! Comments
+|-
+| 32
+| StopHash
+| []byte
+| Block hash for the last filter header returned, for locating the filter headers in the blockchain.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter headers being returned.
+|-
+| Var-Int
+| NumHeaders
+| uint64
+| Hash to stop at.
+|-
+| NumHeaders * 32
+| HeaderHashes
+| [NumHeaders][32]byte
+| Slice of filter headers.
+|}
+
+=== Compact Filters ===
+
+The last set of messages we introduce are for ''fetching'' the compact filters
+themselves. Light clients can SHOULD use these two messages to request a
+compact filter for a particular block hash.
+
+The <code>getcfilter</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| 32
+| BlockHash
+| [32]byte
+| Block hash of the Bitcoin block for which the client wishes to fetch a filter.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter requested.
+|}
+
+The <code>cfilter</code> message MUST be sent in response to a
+<code>getcfilter</code> message for a particular block hash.The
+<code>cfilter</code> message is defined as follows:
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| 32
+| BlockHash
+| [32]byte
+| Block hash of the Bitcoin block for which the filter is being returned.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter being returned.
+|-
+| Var-Int
+| NumFilterBytes
+| uint64
+| A variable length integer encoding the number of bytes of the filter in the following field.
+|-
+| NumFilterBytes
+| FilterBytes
+| [NumFilterBytes]byte
+| The raw compressed compact filter for this block.
+|}
+
+The <code>BlockHash</code> field is included in both messages as this allows
+easily matching requests against responses, as the responses aren't
+self-identifying like block headers are (via own hash).
+
+The parameters <code>N</code> (the number of elements in the filter) and
+<code>P</code> (<code>1 << false_positive_rate</code>) are required by the light
+client in order to properly incrementally decode, query, and validate
+(reconstruct from Bitcoin block) a compact filter. The parameter <code>N</code>
+cannot be known ahead of time, therefore we define the serialization of a
+compact filter of type <code>0x00</code> and <code>0x01</code> as:
+<pre>
+N || raw_filter_bytes
```

Comment:
Why not just put this in the message encoding? If a future new filter type no longer needs the N to be encoded, it can redefine the message encoding then.